deeprankcore/query.py

```diff
@@ -117,6 +125,130 @@ def __repr__(self) -> str:
+        """ Adds new query to the collection of all generated queries.
+            Args:
+                query (Query): must be a Query object, either ProteinProteinInterfaceResidueQuery or SingleResidueVariantAtomicQuery.
+        """
+        self._queries.append(query)
+
+    def export_dict(self, dataset_path: str):
+        """ Exports the colection of all queries to a dictionary file
+            Args:
+                dataset_path (str): the new path where the list of queries be saved to.
+        """
+        with open(dataset_path, "wb") as pkl_file:
+            pickle.dump(self, pkl_file)    
+            
+    @property
+    def queries(self) -> List[Query]:
+        return self._queries
+
+    def __contains__(self, query: Query) -> bool:
+        return query in self._queries
+
+    def __iter__(self) -> Iterator[Query]:
+        return iter(self._queries)
+
+    def _process_one_query(
+        self,
+        prefix: str,
+        feature_names: List[str],
+        query: Query):
+
+        _log.info(f'\nProcess query with process ID {os.getpid()}.')
+
+        # because only one process may access an hdf5 file at the time:
+        output_path = f"{prefix}-{os.getpid()}.hdf5"
+
+        feature_modules = [
+            importlib.import_module('deeprankcore.features.' + name) for name in feature_names]
+
+        graph = query.build(feature_modules)
+
+        graph.write_to_hdf5(output_path)
+
+    def process( # pylint: disable=too-many-locals
+        self, 
+        prefix: Optional[str] = None,
+        process_count: Optional[int] = None,
+        combine_files: bool = True,
+        feature_modules: List[ModuleType] = None
+        ) -> List[str]:
+
+        """
+        Args:
+            prefix: prefix for the output files. ./processed-queries- by default.
+            
+            process_count: how many subprocesses to be run simultaneously.
+                By default takes all available cpu cores.
+
+            combine_files: boolean for combining the hdf5 files generated by the subprocesses.
+                By default, the hdf5 files generated are combined into one, and then deleted.
+
+            feature_modules: list of features' modules used to generate features.
+                Each feature's module must implement the add_features function, and
+                features' modules can be found (or should be placed in case of a custom made feature)
+                in deeprankcore.features folder.
+                If None, all available modules in deeprankcore.features are used to generate the features. 
+                Defaults to None.
+        """
+
+        if process_count is None:
+            # returns the set of CPUs available considering the sched_setaffinity Linux system call,
+            # which limits which CPUs a process and its children can run on.
+            process_count = len(os.sched_getaffinity(0))
+
+        _log.info(f'\nSet of CPU processors available: {process_count}.')
```

Comment:
Should there be an error or warning raised if `process_count` is set higher than the number of available CPUs? I am assuming that that is not possible or at least not a good idea.

Or perhaps a just warning and then defaulted to number of CPUs anyway.

And I guess the log info should state CPUs _used_, not CPUs _available_

Reply:
I added a warning that will be printed in the log file. Also, if the CPUs available are less than `subprocesses`, the latter is defaulted to the number of CPUs available.